### ISSUE TITLE
Make child entities deletable for users with the delete but without the accompanying edit right [5]

### ISSUE THREAD

**Issue by adinaflorea9** [hzi-braunschweig/SORMAS-Project#11557]
### Problem Description

Ticket #11509 introduces the following behavior: users that have delete rights but no edit rights over entities can now click the 'Delete' button.

The following child entities are not covered:
- [x] Prescription
- [x] Treatment 
- [x] Follow-up  
- [x] Clinical course  
- [x] Pathogen Test  

### Proposed Change

Enable the 'Delete' button for the child entities mentioned in the 'Problem description' if the user has the respective delete right, even if they don't have the edit right for the child entity or its parent entity, as long as the entity is in the user's jurisdiction.

### Acceptance Criteria

- a user should be able to delete the child entities for which they have deletion right but no edit right;
- the parent entity needs to be within the users jurisdiction for the user to be able to delete the parent entity or the child entity (jurisdiction based on disease and on responsible region and district).

### Added value/Benefit
Consistency across the app in what concerns the possibility to delete entities over which the user does not have edit right, but has delete right.

### Additional Information

Note: The changes for the Android app will be handled in a separate story: https://github.com/hzi-braunschweig/SORMAS-Project/issues/11892

**Comment by carina29** [comment]
Update related to the Follow-up child entity:
-Will be done for both parent entities -> Cases and Contacts

**Note:**
The isEditable, isDeletable will not depend only on the parent's edit/delete right, children's edit/delete right and the isEditAllowed flag but also on a calculated canEdit boolean flag.

boolean canEdit = VisitOrigin.USER.equals(visit.getOrigin());

**Comment by MateStrysewske** [comment]
@adinaflorea9 I just realised that part of this issue was invalid, but for some reason I missed that earlier. We can't make the delete right of child entities depending on the delete right of the parent entity; users definitely have to be able to delete e.g. prescriptions and treatments if they're not allowed to delete whole cases. I've updated the title and issue description accordingly.

**Comment by carina29** [comment]
**Note:** 
If editing or deleting is allowed, the heading of the modal window will contain "Edit ...." keyword. 
Otherwise "View ..." keyword will be displayed.  (For Prescription we already had a heading when only view right exists - is "Prescription" not "View prescription" )

![image](https://user-images.githubusercontent.com/47103965/233938544-1742a357-01e0-4ec0-a140-377aefd32384.png)

![image](https://user-images.githubusercontent.com/47103965/233939082-741ff6ba-58e0-4f60-b227-904ff73c9e7c.png)

**Comment by MateStrysewske** [comment]
@carina29 Make sure that it's "View visit", not "Visit view" :)

**Comment by carina29** [comment]
**Update:**

Edit on child will be allowed if: isEditAllowed flag is true, edit on Parent is true and + edit on Child is true
Delete on child will be allowed if : isEditAllowed flag is true, delete on Child is true

I've used this table for testing:
![image](https://user-images.githubusercontent.com/47103965/234015775-4e347b8a-7df3-4ec9-af33-1f2b1e032d3e.png)

**Comment by carina29** [comment]
**Update:**
Related to the above comments, all the heading titles will have the following format:
1) View {child entity name} (if there is not edit or delete right)
2) Edit {child entity name} (if there is edit or delete rights or both)

**Comment by adinaflorea9** [comment]
I am reopening the ticket due to the following - a user can not delete only a prescription if the prescription also has a treatment associated. 

Steps to reproduce:
1. Logged in with an Admin+NatUser, open an existing Covid-19 case;
2. Create a new prescription for the case;
3. In the prescription section click on 'Create treatment' (the button that is on the same line as the prescription created at step 2);
4. Save the new treatment;
5. Log in the app with a user that has the following rights:
![image](https://user-images.githubusercontent.com/96909231/236202372-e2433702-626e-403d-a785-52032cbd56e7.png)
6. Open the prescription created at step 2 and click 'Delete';
7. In the 'Confirm deletion' pop-up click 'Yes';
8. In the 'Delete prescriptions with treatments' pop-up click 'Just Prescriptions'.
Actual behavior: An error appears:
![image](https://user-images.githubusercontent.com/96909231/236203459-519b0476-cd46-44f6-b1bc-9fb4acc2cb67.png)

**Expected behavior:** The user should be able to successfully delete the prescription alone.

Logs from the local machine:
[Error delete prescription without treatment .txt](https://github.com/hzi-braunschweig/SORMAS-Project/files/11397419/Error.delete.prescription.without.treatment.txt)

**Comment by MateStrysewske** [comment]
Actually, the user should not be able to delete that prescription because it's referenced by the treatments that have been created based on it. The issue here is rather that the error message displayed is simply wrong. What should be displayed is the following: "This prescription can't be deleted because it is referred to in at least one treatment. Please delete all treatments referring this prescription first."

**Comment by adinaflorea9** [comment]
@MateStrysewske but the same steps on release instances allow for the deletion of prescriptions alone, even if they have a treatment associated. Was that never intended to be the behavior? As I though that that is why we have the option to delete only the prescription and that button specific in that pop-up

**Comment by MateStrysewske** [comment]
@adinaflorea9 That is very strange and shouldn't be possible technically because the prescription should be referenced by the treatment in the database, so trying to delete the prescription is expected to lead to an error; I don't think deleting it cascades either, but you could try that out; on the release instance, when you create a prescription, create a treatment based on that prescription and then delete the prescription, is the treatment also deleted afterwards?

**Comment by MateStrysewske** [comment]
@adinaflorea9 Nevermind, you're correct, we specifically have a pop-up that asks the user whether they want to delete only the prescription or also the treatments. Will investigate this further.

**Comment by adinaflorea9** [comment]
No, the treatment remains in that scenario but only if you select the following button in the pop-up:
![image](https://user-images.githubusercontent.com/96909231/236417745-31d27985-84ec-41c0-909a-d8a7f85e4cb5.png)

So if we have a pop-up that gives the user the option to choose whether they want to delete them both at the same time or to delete only the prescription, this is why I thought that this is the expected behavior.

When the user also has treatment edit rights the process works as expected even with Carina's changes. The logs state that the user needs to have that right I presume because somehow you are sort of editing a treatment at the time when you delete the prescription associated as you are breaking that link they used to have.

**Comment by MateStrysewske** [comment]
@adinaflorea9 Yes, exactly, we need to edit the treatment because we have to remove the prescription from it. 

@carina29 I'd say the solution is to add a specific method called `removePrescription` to the TreatmentFacade that is annotated with the user rights TREATMENT_EDIT and PRESCRIPTION_DELETE.

**Comment by adinaflorea9** [comment]
#11932 is believed to be linked to this ticket. 

For a user that has the right to 'Edit existing clinical visits' but does not have 'Edit existing cases' they can not successfully edit an existing clinical visit. The logs displayed on the local machine state that the missing rights are CASE_CREATE and CASE_EDTI, bunt once case edit is given to the user they can edit the clinical visits, so it is also unclear why case create is mentioned in the logs.

Additionally, with a similar combination of rights, such as edit existing treatments, or edit existing follow-up visits and the missing case edit right, the user can successfully edit the treatment or the follow-up visit over which they have rights for.

Therefore, for consistency, they should also be able to edit existing clinical visit when they have the right assigned.

Reproduction Steps
1. Create a user with the following rights:
![image](https://user-images.githubusercontent.com/96909231/236448919-bd9e4960-04df-4b50-be10-034fa803fc1d.png)

2. With a regular Admin+NatUser log in and create a clinical visit for a case;
3. Log in with a user that has the rights from step 1 and open the case;
4. Open the existing clinical visit, edit something on the page and then click 'Save'.

Logs from the local machine server:
[Error clinical visit edit .txt](https://github.com/hzi-braunschweig/SORMAS-Project/files/11405826/Error.clinical.visit.edit.txt)


Expected Behavior
The user should be able to successfully edit the clinical visit.

**Comment by adinaflorea9** [comment]
Users that are limited by a disease can delete the child entities associated with parent entities of other diseases (as long as they have the delete rights for the respective entity).

The disease jurisdiction should be taken into account here as well as mentioned in the acceptance criteria.

**Comment by carina29** [comment]
Related to the finding No1:
@adinaflorea9 @MateStrysewske I've checked the code and the error from the log files.
We already have a method in TreatmentFacadeEjb.java unlinkPrescriptionFromTreatment which right now only the TREATMENT_EDIT rights have.
![image](https://github.com/hzi-braunschweig/SORMAS-Project/assets/47103965/4151414c-05a9-48be-b78e-20fc87279e1a)

The error from the logs is thrown because the logged in user does not have a TREATMENT_EDIT.
I've added the PRESCRIPTION_DELETE right to the unlinkPrescriptionFromTreatments method and also checked the usages of this method: it's used only when the prescriptions are deleted ( in single or bulk mode) so it looks that the PRESCRIPTION_DELETE would be enough for this scenario and TREATMENT_EDIT is redundant. 
Right now I've left both rights added to the method. 

Related to finding No2:
The problem there is that when a user does not have edit right ( a user has edit right if there is an edit right on the parent + an edit right on the child entity) the form should be readonly.
I've done a refactoring on the initial PR which did not checked both edit rights ( on parent and child) that's why Adina was able to edit the form.
I've talked with @adinaflorea9 about this and corrected the bug.
